src/System.Management.Automation/engine/CommandCompletion/ExtensibleCompletion.cs

```diff
@@ -102,7 +102,127 @@ IEnumerable<CompletionResult> CompleteArgument(
             string parameterName,
             string wordToComplete,
             CommandAst commandAst,
-            IDictionary fakeBoundParameters);
+            IDictionary fakeBoundParameters)
+        {
+            CommandName = commandName;
+            ParameterName = parameterName;
+            WordToComplete = wordToComplete;
+            CommandAst = commandAst;
+            FakeBoundParameters = fakeBoundParameters;
+            return ShouldComplete ? GetMatchingResults(wordToComplete) : [];
+        }
+
+        /// <summary>
+        /// Gets the name of the command that needs argument completion.
+        /// </summary>
+        protected static string? CommandName { get; private set; }
+
+        /// <summary>
+        /// Gets the name of the parameter that needs argument completion.
+        /// </summary>
+        protected static string? ParameterName { get; private set; }
+
+        /// <summary>
+        /// Gets the word being completed.
+        /// </summary>
+        protected static string? WordToComplete { get; private set; }
+
+        /// <summary>
+        /// Gets the command abstract syntax tree (AST).
+        /// </summary>
+        protected static CommandAst? CommandAst { get; private set; }
+
+        /// <summary>
+        /// Gets the fake bound parameters similar to $PSBoundParameters.
+        /// </summary>
+        protected static IDictionary? FakeBoundParameters { get; private set; }
+
+        /// <summary>
+        /// Gets value indicating whether to perform completion.
+        /// </summary>
+        protected bool ShouldComplete => true;
+
+        /// <summary>
+        /// Gets the type of the completion result.
+        /// </summary>
+        protected CompletionResultType CompletionResultType => CompletionResultType.Text;
+
+        /// <summary>
+        /// Gets the mapping function for tooltips.
+        /// </summary>
+        protected Func<string, string>? ToolTipMapping => null;
+
+        /// <summary>
+        /// Gets the mapping function for list item texts.
+        /// </summary>
+        protected Func<string, string>? ListItemTextMapping => null;
+
+        /// <summary>
+        /// Gets value indicating whether to escape globbing paths.
+        /// </summary>
+        protected bool EscapeGlobbingPath => false;
+
+        /// <summary>
+        /// Gets the possible completion values.
+        /// </summary>
+        protected IEnumerable<string> PossibleCompletionValues => [];
```

Comment:
Made sense to just have all these as `get;` properties. Completers just implement their own properties if behaviour needs to change.